Added docstring to @fastmath macro.

NHDaly opened the pull request:
Adds docstring to @fastmath macro that explains a bit about what it does, and includes examples how to use it.

Pulled the detailed list of effects from the top of the file.

#doc

NHDaly committed:
Added docstring to @fastmath macro.

Adds docstring to @fastmath macro that explains a bit about what it does, and includes examples how to use it.

yuyichao commented on base/fastmath.jl:
```diff
@@ -124,6 +124,32 @@ function make_fastmath(symb::Symbol)
+- nsz:  No Signed Zeros - Allow optimizations to treat the sign of a
+        zero argument or result as insignificant.
+- arcp: Allow Reciprocal - Allow optimizations to use the reciprocal
+        of an argument rather than perform division.
```
Assuming you are copying from langref, the transformation allowed is much broader than this list.

yuyichao replied:
Ah, I see that you pull the list from the comment, in that case, maybe you should use http://llvm.org/docs/LangRef.html#fast-math-flags

NHDaly replied:
Oh, thanks! I've updated with a second commit. I can squash before merging.

Does this look better? I added a link to the llvm docs (and added one at the top of the file), and I listed the missing flags.

ararslan replied:
No need to manually squash; we can squash the commits on merge.

NHDaly committed:
Updated per yuyichao's comments

Added link to  [LLVM Fast-Math flags](http://llvm.org/docs/LangRef.html#fast-math-flags)

yuyichao reviewed (approved):
I'd prefer shrinking this a little (i.e. `@fastmath` allows all sorts of transformation and then mention the few specific ones as special cases) but this looks correct and complete enough to me....

yuyichao commented on base/fastmath.jl:
```diff
@@ -124,6 +125,38 @@ function make_fastmath(symb::Symbol)
 make_fastmath(expr) = expr
 
+"""
+@fastmath expr
```
Actually, this should be a code block (indent it) and this should be added to the doc/stdlib/

NHDaly replied:
>Actually, this should be a code block (indent it) 

Just indent everything inside the triple quotes, you mean?

> and this should be added to the doc/stdlib/

Sorry, I don't understand what you mean. Like, there should be a page for this in the official docs? Like a new markdown file?

yuyichao replied:
> Just indent everything inside the triple quotes, you mean?

No, just the signature, see [the doc for other macros/functions](https://github.com/JuliaLang/julia/blob/d3db31200c09a99ad9f7cbc019dd0aa56c6944a7/base/essentials.jl#L200)

> added to the doc/stdlib/

See https://github.com/JuliaLang/julia/blob/d3db31200c09a99ad9f7cbc019dd0aa56c6944a7/CONTRIBUTING.md#improving-documentation There are examples there that you can follow. And you can decide which file to put it in...

NHDaly replied:
>No, just the signature, see the doc for other macros/functions

Gotcha.  Done. thanks!

See [CONTRIBUTING.md#improving-documentation](https://github.com/JuliaLang/julia/blob/d3db31200c09a99ad9f7cbc019dd0aa56c6944a7/CONTRIBUTING.md#improving-documentation)

Okay thanks. I made those changes, and I'm running `make docs` now. I'll post again when that's done.

NHDaly replied:
oh noes! Jk i can't build on this computer; i don't have all the dependencies installed. I'll push this tomorrow. Thanks! :)

yuyichao replied:
Note that `make docs` is only needed for testing. There shouldn't be anything that you need to push after that.

NHDaly replied:
Gotcha. I have verified the docs build and look correct.

NHDaly committed:
Indented docstring first line

NHDaly committed:
added @fastmath to math.md @doc

ararslan commented:
Thanks for the contribution! For future reference, it's a good idea to reference macros in commit messages as e.g. `at-fastmath rather than `@fastmath`. In the latter case, GitHub will send a notification to the user called fastmath. Whoever squash-merges this can edit the commit message here appropriately.

NHDaly commented:
> I'd prefer shrinking this a little (i.e. @fastmath allows all sorts of transformation and then mention the few specific ones as special cases) but this looks correct and complete enough to me....

@yuyichao I agree. What about just dropping everything after the definition, and dropping `fast` since it's so vague, so it looks like this:

```
"""
    @fastmath expr

Executes a transformed version of the expression, which calls
functions that may violate strict IEEE semantics. This allows
the fastest possible operation, but results are undefined.

This sets the [LLVM Fast-Math
flags](http://llvm.org/docs/LangRef.html#fast-math-flags), including
the following transformations:
- nnan: No NaNs
- ninf: No Infs
- nsz:  No Signed Zeros
- arcp: Allow Reciprocal
- contract: Allow floating-point contraction.

# Examples
```jldoctest
julia> @fastmath 1+2
3

julia> @fastmath(sin(3))
0.1411200080598672
```
"""
```

NHDaly commented:
> For future reference, it's a good idea to reference macros in commit messages as e.g. `at-fastmath` rather than `@fastmath`. In the latter case, GitHub will send a notification to the user called fastmath. Whoever squash-merges this can edit the commit message here appropriately.

:D haha that makes a lot of sense. Thanks!! I'll let you do the change so my rebase doesn't make the PR harder to read.

Cheers!
~Nathan

yuyichao commented:
> dropping fast since it's so vague,

I'll not drop the `fast` description though it'll be better to drop those flag names altogether, maybe only with a mention that this corresponds to the llvm `fast` flag. There are many more transformations allowed for `fast` that's not included in those categories.

NHDaly committed:
added link to performance annotations, removed bulleted details

NHDaly commented:
> I'll not drop the fast description though it'll be better to drop those flag names altogether, maybe only with a mention that this corresponds to the llvm fast flag. There are many more transformations allowed for fast that's not included in those categories.

Okay, I've pushed a new version that removes the detailed bullets and instead links to the following page in the docs:
https://docs.julialang.org/en/stable/manual/performance-tips#performance-annotations

Is there a better way to link to docs, such that it will keep you in the correct, corresponding version? (ie don't always link to `stable`?)

fredrikekre commented:
> Is there a better way to link to docs

You can use `@ref` if you add an `@id` to the section header. Something like this:
```patch
diff --git a/base/fastmath.jl b/base/fastmath.jl
index f6be79a..6f925e4 100644
--- a/base/fastmath.jl
+++ b/base/fastmath.jl
@@ -133,7 +133,10 @@ functions that may violate strict IEEE semantics. This allows
 the fastest possible operation, but results are undefined.
 
 This sets the [LLVM Fast-Math
-flags](http://llvm.org/docs/LangRef.html#fast-math-flags), and corresponds to the `-ffast-math` option in clang. See [the notes on performance annotations](https://docs.julialang.org/en/stable/manual/performance-tips#performance-annotations) for more details.
+flags](http://llvm.org/docs/LangRef.html#fast-math-flags), and corresponds to the
+`-ffast-math` option in clang.
+See [the notes on performance annotations](@ref man-performance-annotations) for
+more details.
 
 # Examples
 ```jldoctest
diff --git a/doc/src/manual/performance-tips.md b/doc/src/manual/performance-tips.md
index d171cbf..02f609b 100644
--- a/doc/src/manual/performance-tips.md
+++ b/doc/src/manual/performance-tips.md
@@ -1082,7 +1082,7 @@ These are some minor points that might help in tight inner loops.
   * Use [`div(x,y)`](@ref) for truncating division of integers instead of [`trunc(x/y)`](@ref), [`fld(x,y)`](@ref)
     instead of [`floor(x/y)`](@ref), and [`cld(x,y)`](@ref) instead of [`ceil(x/y)`](@ref).
 
-## Performance Annotations
+## [Performance Annotations](@id man-performance-annotations)
 
 Sometimes you can enable better optimization by promising certain program properties.

```

StefanKarpinski commented on base/fastmath.jl:
```diff
@@ -124,6 +125,25 @@ function make_fastmath(symb::Symbol)
+the fastest possible operation, but results are undefined.
+
+This sets the [LLVM Fast-Math
+flags](http://llvm.org/docs/LangRef.html#fast-math-flags), and corresponds to the `-ffast-math` option in clang. See [the notes on performance annotations](https://docs.julialang.org/en/stable/manual/performance-tips#performance-annotations) for more details.
```
wrap 92 please

NHDaly replied:
Oops. Done, thanks.

NHDaly committed:
fixed intra-doc link && wrap 92.

NHDaly committed:
added "be careful" warning as well.

NHDaly commented:
PTAL: I think this is all set, and the docs build correctly. Thanks again for your reviewing!

> You can use @ref if you add an @id to the section header. Something like this:

Thanks, done.

yuyichao commented:
Be careful to code quote all `@`s that are not intended to mention people, even in text quote.....

I did not know that `ref` and `id` are real github users.... (with `id` actually seems to be active)

NHDaly commented:
Haha wow that's hard. I'll be sure to be careful. I didn't do it again, did I?

yuyichao commented:
> I didn't do it again, did I?

I edited the comment to fix it.

ararslan reviewed (approved):
LGTM. Thanks for the contribution!

Sacha0 commented on base/fastmath.jl:
```diff
@@ -124,6 +125,27 @@ function make_fastmath(symb::Symbol)
+"""
+    @fastmath expr
+
+Executes a transformed version of the expression, which calls functions that
```
Perhaps `Executes` -> `Execute` in keeping with the documentation guidelines? :)

NHDaly replied:
Done, thanks! Sorry about that.

Sacha0 reviewed (approved):
lgtm! Thanks @NHDaly! :)

NHDaly committed:
s/Executes/Execute/g

Sacha0 commented:
Thanks for the great contribution @NHDaly ! :)

fastmath commented:
Thank you for documenting me.
